File: frontend/src/store/highlights.ts
import { create } from 'zustand'
import apiClient from '../apiClient'
import type { Highlight } from '../types'

interface HighlightCreate {
  selected_text: string
  prefix_context: string
  suffix_context: string
  color: string
  note?: string
}

interface HighlightUpdate {
  color?: string
  note?: string | null
  thesis_section?: string | null
}

interface HighlightsState {
  /** Highlights keyed by article_id */
  highlights: Record<number, Highlight[]>

  fetchHighlights: (articleId: number) => Promise<void>
  createHighlight: (articleId: number, data: HighlightCreate) => Promise<Highlight>
  updateHighlight: (articleId: number, id: number, data: HighlightUpdate) => Promise<void>
  deleteHighlight: (articleId: number, id: number) => Promise<void>
  patchHighlight: (id: number, data: HighlightUpdate) => Promise<void>
}

export const useHighlightsStore = create<HighlightsState>((set, get) => ({
  highlights: {},

  async fetchHighlights(articleId) {
    try {
      const data = await apiClient.get<Highlight[]>(`/api/articles/${articleId}/highlights`)
      set((s) => ({ highlights: { ...s.highlights, [articleId]: data } }))
    } catch {
      // Non-fatal: silently skip if article has no highlights or request fails
    }
  },

  async createHighlight(articleId, data) {
    const created = await apiClient.post<Highlight>(`/api/articles/${articleId}/highlights`, data)
    set((s) => ({
      highlights: {
        ...s.highlights,
        [articleId]: [...(s.highlights[articleId] ?? []), created],
      },
    }))
    return created
  },

  async updateHighlight(articleId, id, data) {
    try {
      const updated = await apiClient.put<Highlight>(`/api/articles/${articleId}/highlights/${id}`, data)
      set((s) => ({
        highlights: {
          ...s.highlights,
          [articleId]: (s.highlights[articleId] ?? []).map((h) =>
            h.id === id ? updated : h
          ),
        },
      }))
    } catch {
      // Non-fatal
    }
  },

  async deleteHighlight(articleId, id) {
    try {
      await apiClient.del(`/api/articles/${articleId}/highlights/${id}`)
      set((s) => ({
        highlights: {
          ...s.highlights,
          [articleId]: (s.highlights[articleId] ?? []).filter((h) => h.id !== id),
        },
      }))
    } catch {
      // Non-fatal
    }
  },

  async patchHighlight(id, data) {
    try {
      const updated = await apiClient.patch<Highlight>(`/api/highlights/${id}`, data)
      set((s) => {
        const next = { ...s.highlights }
        for (const aid of Object.keys(next)) {
          const aidNum = Number(aid)
          next[aidNum] = next[aidNum].map((h) => (h.id === id ? updated : h))
        }
        return { highlights: next }
      })
    } catch {
      // Non-fatal
    }
  },
}))

File: backend/api/routers/highlights.py
from datetime import datetime, timezone
from typing import List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from auth import require_session
from database import Article, Highlight, get_db, get_valid_thesis_sections
from models import BulkUpdateHighlightsRequest, HighlightCreate, HighlightOut, HighlightUpdate

router = APIRouter()
_auth = Depends(require_session)


@router.get("/api/articles/{article_id}/highlights", response_model=List[HighlightOut])
async def list_highlights(
    article_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_session),
):
    article = db.query(Article).filter(Article.id == article_id).first()
    if not article:
        raise HTTPException(status_code=404, detail="Article not found")
    highlights = (
        db.query(Highlight)
        .filter(Highlight.article_id == article_id, Highlight.user_id == current_user["id"])
        .order_by(Highlight.created_at)
        .all()
    )
    return [HighlightOut.model_validate(h) for h in highlights]


@router.post("/api/articles/{article_id}/highlights", response_model=HighlightOut, status_code=201)
async def create_highlight(
    article_id: int,
    body: HighlightCreate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_session),
):
    article = db.query(Article).filter(Article.id == article_id).first()
    if not article:
        raise HTTPException(status_code=404, detail="Article not found")
    highlight = Highlight(
        article_id=article_id,
        user_id=current_user["id"],
        selected_text=body.selected_text,
        prefix_context=body.prefix_context,
        suffix_context=body.suffix_context,
        color=body.color,
        note=body.note,
        created_at=datetime.now(timezone.utc),
    )
    db.add(highlight)
    db.commit()
    db.refresh(highlight)
    return HighlightOut.model_validate(highlight)


@router.put("/api/articles/{article_id}/highlights/{highlight_id}", response_model=HighlightOut)
async def update_highlight(
    article_id: int,
    highlight_id: int,
    body: HighlightUpdate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_session),
):
    highlight = (
        db.query(Highlight)
        .filter(
            Highlight.id == highlight_id,
            Highlight.article_id == article_id,
            Highlight.user_id == current_user["id"],
        )
        .first()
    )
    if not highlight:
        raise HTTPException(status_code=404, detail="Highlight not found")
    if body.color is not None:
        highlight.color = body.color
    if body.note is not None:
        highlight.note = body.note
    if body.thesis_section is not None:
        valid = get_valid_thesis_sections(db, current_user["id"])
        if body.thesis_section not in valid:
            raise HTTPException(
                status_code=422,
                detail=f"thesis_section must be one of {sorted(valid)}",
            )
        highlight.thesis_section = body.thesis_section
    db.commit()
    db.refresh(highlight)
    return HighlightOut.model_validate(highlight)


@router.patch("/api/highlights/{highlight_id}", response_model=HighlightOut)
async def patch_highlight(
    highlight_id: int,
    body: HighlightUpdate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_session),
):
    """Partial update of a highlight — only the fields present in the body are updated."""
    h = (
        db.query(Highlight)
        .filter(Highlight.id == highlight_id, Highlight.user_id == current_user["id"])
        .first()
    )
    if not h:
        raise HTTPException(status_code=404, detail="Highlight not found")
    if body.color is not None:
        h.color = body.color
    if body.note is not None:
        h.note = body.note
    if body.thesis_section is not None:
        valid = get_valid_thesis_sections(db, current_user["id"])
        if body.thesis_section not in valid:
            raise HTTPException(
                status_code=422,
                detail=f"thesis_section must be one of {sorted(valid)}",
            )
        h.thesis_section = body.thesis_section
    db.commit()
    return HighlightOut.model_validate(h)


@router.post("/api/research/highlights/bulk-update")
async def bulk_update_highlights(
    body: BulkUpdateHighlightsRequest,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_session),
):
    """Update thesis_section for multiple highlights at once."""
    valid = get_valid_thesis_sections(db, current_user["id"])
    if body.thesis_section not in valid:
        raise HTTPException(
            status_code=422,
            detail=f"thesis_section must be one of {sorted(valid)}",
        )
    updated = (
        db.query(Highlight)
        .filter(
            Highlight.id.in_(body.highlight_ids),
            Highlight.user_id == current_user["id"],
        )
        .update({Highlight.thesis_section: body.thesis_section}, synchronize_session=False)
    )
    db.commit()
    return {"status": "ok", "updated": updated}


@router.delete("/api/articles/{article_id}/highlights/{highlight_id}")
async def delete_highlight(
    article_id: int,
    highlight_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_session),
):
    highlight = (
        db.query(Highlight)
        .filter(
            Highlight.id == highlight_id,
            Highlight.article_id == article_id,
            Highlight.user_id == current_user["id"],
        )
        .first()
    )
    if not highlight:
        raise HTTPException(status_code=404, detail="Highlight not found")
    db.delete(highlight)
    db.commit()
    return {"status": "ok"}
